packages/@aws-cdk/tmp-toolkit-helpers/src/api/refactoring/index.ts

```diff
@@ -0,0 +1,415 @@
+  const resources = template.Resources || {};
+  const graph: Record<string, Set<string>> = {};
+  const reverseGraph: Record<string, Set<string>> = {};
+
+  // 1. Build adjacency lists
+  for (const id of Object.keys(resources)) {
+    graph[id] = new Set();
+    reverseGraph[id] = new Set();
+  }
+
+  // 2. Detect dependencies by searching for Ref/Fn::GetAtt
+  const findDependencies = (value: any): string[] => {
+    if (!value || typeof value !== 'object') return [];
+    if (Array.isArray(value)) {
+      return value.flatMap(findDependencies);
+    }
+    if ('Ref' in value) {
+      return [value.Ref];
+    }
+    if ('Fn::GetAtt' in value) {
+      const refTarget = Array.isArray(value['Fn::GetAtt']) ? value['Fn::GetAtt'][0] : value['Fn::GetAtt'].split('.')[0];
+      return [refTarget];
+    }
+    return Object.values(value).flatMap(findDependencies);
+  };
+
+  for (const [id, res] of Object.entries(resources)) {
+    const deps = findDependencies(res.Properties || {});
+    for (const dep of deps) {
+      if (dep in resources && dep !== id) {
+        graph[id].add(dep);
+        reverseGraph[dep].add(id);
+      }
+    }
+  }
+
+  // 3. Topological sort
+  const inDegree = Object.keys(graph).reduce((acc, k) => {
+    acc[k] = graph[k].size;
+    return acc;
+  }, {} as Record<string, number>);
+
+  const queue = Object.keys(inDegree).filter((k) => inDegree[k] === 0);
+  const order: string[] = [];
+
+  while (queue.length > 0) {
+    const node = queue.shift()!;
+    order.push(node);
+    for (const nxt of reverseGraph[node]) {
+      inDegree[nxt]--;
+      if (inDegree[nxt] === 0) {
+        queue.push(nxt);
+      }
+    }
+  }
+
+  // 4. Compute digests in sorted order
+  const result: Record<string, string> = {};
+  for (const id of order) {
+    const resource = resources[id];
+    const depDigests = Array.from(graph[id]).map((d) => result[d]);
+    const propsWithoutRefs = JSON.stringify(stripReferences(stripConstructPath(resource.Properties)));
+    const toHash = resource.Type + propsWithoutRefs + depDigests.join('');
+    result[id] = crypto.createHash('sha256').update(toHash).digest('hex');
+  }
+
+  return result;
+}
+
+function hashObject(obj: any): string {
+  const hash = createHash('sha256');
+
+  function addToHash(value: any) {
+    if (typeof value === 'object') {
+      if (value == null) {
+        addToHash('null');
+      } else if (Array.isArray(value)) {
+        value.forEach(addToHash);
+      } else {
+        Object.keys(value)
+          .sort()
+          .forEach((key) => {
+            hash.update(key);
+            addToHash(value[key]);
+          });
+      }
+    } else {
+      hash.update(typeof value + value.toString());
+    }
+  }
+
+  addToHash(obj);
+  return hash.digest('hex');
+}
+
+/**
+ * Removes sub-properties containing Ref or Fn::GetAtt to avoid hashing
+ * references themselves but keeps the property structure.
```

Comment:
Surely though, it matters WHICH object we reference, not just that we reference SOME object?

Wouldn't it be better to replace the logical ID with the digest of the referenced source?

Currently:

```
{ "Function": { "Ref": "MyFunction" } }
-->
{ "Function": { __cloud_ref__: "Ref" } }


{ "Function": { "Fn::GetAtt": ["MyFunction", "Arn"] } }
-->
{ "Function": { __cloud_ref__: "Fn::GetAtt" } }
```

Instead:

```
{ "Function": { "Ref": "MyFunction" } }
-->
{ "Function": { "Ref": "baf399b3cdafe24b4b6e1b0674bb59ae" } }


{ "Function": { "Fn::GetAtt": ["MyFunction", "Arn"] } }
-->
{ "Function": { "Fn::GetAtt": ["baf399b3cdafe24b4b6e1b0674bb59ae", Arn" ] } }
```

?

Reply:
EDIT: oh I see, you mix the dependencies in elsewhere